lib/commons/color/get-foreground-color.js

```diff
@@ -83,26 +90,75 @@ function getStrokeColor(nodeStyle, { textStrokeEmMin = 0 }) {
   return new Color().parseString(strokeColor);
 }
 
-function getOpacity(node, nodeStyle) {
-  if (!node) {
-    return 1;
-  }
+/**
+ * Blend a foreground color into the background stacking context, taking into account opacity at each step.
+ * @param {Color} fgColor
+ * @param {Object} context - The nodes stacking context
+ * @param {Object[]} stackingContexts - Array of all stacking contexts
+ * @return {Color}
+ */
+function calculateBlendedForegroundColor(fgColor, context, stackingContexts) {
+  while (context) {
+    // find the nearest ancestor that has opacity < 1
+    if (context.opacity === 1 && context.ancestor) {
+      context = context.ancestor;
+      continue;
+    }
 
-  const vNode = getNodeFromTree(node);
-  if (vNode && vNode._opacity !== undefined && vNode._opacity !== null) {
-    return vNode._opacity;
-  }
+    fgColor.alpha *= context.opacity;
+
+    // when blending the foreground color to a background color with opacity,
+    // we ignore the background color of the node itself and instead blend
+    // with the stack behind it
+    let stack = context.ancestor?.descendants || stackingContexts;
+    if (context.opacity !== 1) {
+      stack = stack.slice(0, stack.indexOf(context));
+    }
 
-  nodeStyle ??= window.getComputedStyle(node);
-  const opacity = nodeStyle.getPropertyValue('opacity');
-  const finalOpacity = opacity * getOpacity(node.parentElement);
+    const bgColors = stack.map(stackingContextToColor);
 
-  // cache the results of the getOpacity check on the parent tree
-  // so we don't have to look at the parent tree again for all its
-  // descendants
-  if (vNode) {
-    vNode._opacity = finalOpacity;
+    if (!bgColors.length) {
+      context = context.ancestor;
+      continue;
+    }
+
+    const bgColor = bgColors.reduce(
+      (backdrop, source) => {
+        return flattenColors(
+          source.color,
+          backdrop.color instanceof Color ? backdrop.color : backdrop
+        );
+      },
+      {
+        color: new Color(0, 0, 0, 0),
+        blendMode: 'normal'
+      }
+    );
+
+    fgColor = flattenColors(fgColor, bgColor);
+
+    context = context.ancestor;
   }
 
-  return finalOpacity;
+  return fgColor;
+}
+
+/**
+ * Find the stacking context that belongs to the passed in node
+ * @param {Object} contexts - Array of stacking contexts
+ * @param {Element} node
+ * @returns {Object}
+ */
+function findContext(contexts, node) {
+  for (let i = 0; i < contexts.length; i++) {
+    const context = contexts[i];
```

Comment:
Is there a reason not to do this?

```suggestion
  for (const context of contexts) {
```